lib/staccato/v4/tracker.rb

```diff
@@ -1,155 +1,334 @@
-      Staccato::V4.ga_collection_uri
+    # (see #post)
+    def track
+      post
     end
 
-    def validation_uri
-      Staccato::V4.validation_uri
+    # Remove the first `count` events from the {#events} array.
+    #
+    # Related: {#clear}
+    #
+    # @return self
+    def trim(count: 25)
+      @events.slice!(0, count)
+      self
     end
 
-    private
-
-    # @private
-    def params
-      {
-        measurement_id: @measurement_id,
-        api_secret: @api_secret
-      }
+    # Remove the first `count` events from the {#events} array.
+    #
+    # Related: {#trim}
+    #
+    # @return self
+    def clear
+      @events.clear
+      self
     end
 
-    # @private
-    def body
-      Staccato::V4.encode_body(client_id, events.map {|event| event.as_json})
+    # Post the first batch_size {#events} to the Measurement Protocol API
+    # Validation Server endpoint.
+    #
+    # Note that, unlike {#track}, the {#events} array will not be automatically
+    # cleared.
+    def validate!
+      self.class.validation_adapter.post_with_body(query_params, body)
     end
 
-    # @private
-    def single_adapter?
-      adapters.length == 1
+    def query_params
+      { measurement_id: measurement_id, api_secret: api_secret }
     end
 
-    # @private
-    def post
-      single_adapter? ? post_first : post_all
+    # n.b. only the first batch_size events are included
+    def as_json(batch_size: 25)
+      {
+        client_id: client_id,
+        user_id: user_id,
+        user_properties: user_properties,
+        non_personalized_ads: non_personalized_ads,
+        timestamp_micros: timestamp_micros,
+        events: Staccato::V4.encode_params_json(@events[0, batch_size])
+      }.compact.reject {|k, v| v.nil? || v.respond_to?(:empty?) && v.empty? }
     end
 
-    # @private
-    def post_first
-      adapters.first.post_with_body(params, body)
+    def adapters
+      @adapters << self.class.default_adapter if @adapters.empty?
+      @adapters
     end
 
-    # @private
-    def post_all
-      adapters.map {|adapter| adapter.post_with_body(params, body)}
-    end
+    private
 
-    # @private
-    def adapters
-      @adapters.empty? ? [default_adapter] : @adapters
+    def body(batch_size:)
+      JSON.generate(as_json(batch_size: batch_size))
     end
 
-    # @private
-    def default_adapter
-      @default_adapter ||= Staccato.default_adapter.new(default_uri)
+    def post_batch(batch_size:, adapters: self.adapters)
+      if adapters.length == 1
+        adapters.first.post_with_body(
+          query_params,
+          body(batch_size: batch_size)
+        )
+      else
+        query_params = self.query_params
+        body         = body(batch_size: batch_size)
+        adapters.map {|adapter| adapter.post_with_body(query_params, body)}
+      end
     end
   end
 
   # A tracker which does no tracking
   #   Useful in testing
-  class NoopTracker
+  class NoopTracker < Tracker
     attr_accessor :hit_defaults
```

Comment:
I probably should have removed this as it's not a hit anymore, just an event. Oops!